gio/src/list_model.rs

```diff
@@ -19,15 +31,149 @@ impl<T: IsA<ListModel>> ListModelExtManual for T {
         }
         res
     }
+    fn iter<LT: IsA<glib::Object>>(&self) -> Result<ListModelIter<LT>, &Self>
+    where
+        Self: Sized + Clone,
+    {
+        if !self.item_type().is_a(LT::static_type()) {
+            return Err(self);
+        }
+
+        let len = self.n_items();
+        let changed = Rc::new(Cell::new(false));
+
+        let changed_clone = changed.clone();
+        let signal_id = Cell::new(Some(self.connect_items_changed(move |_, pos, _, _| {
+            changed_clone.replace(pos < len);
+        })));
+
+        Ok(ListModelIter {
+            ty: Default::default(),
+            i: 0,
+            reverse_pos: len,
+            model: self.clone().upcast(),
+            changed,
+            signal_id,
+        })
+    }
 }
 
-impl std::iter::IntoIterator for ListModel {
-    type Item = glib::Object;
-    type IntoIter = std::vec::IntoIter<glib::Object>;
+#[derive(thiserror::Error, Debug, PartialEq)]
+#[error("the list model was mutated during iteration")]
+pub struct ListModelMutatedDuringIter;
+
+// rustdoc-stripper-ignore-next
+/// Iterator of `ListModel`'s items.
+/// This iterator will always give `n = initial_model.n_items()` items, even if the `ListModel`
+/// is mutated during iteration.
+/// If the internal `ListModel` gets mutated, the iterator
+/// will return `Some(Err(...))` for the remaining items.
+/// Mutations to the `ListModel` in position >= `initial_model.n_items()` are allowed.
+pub struct ListModelIter<T: IsA<glib::Object>> {
+    ty: PhantomData<T>,
+    i: u32,
+    // it's > i when valid
+    reverse_pos: u32,
+    model: ListModel,
+    changed: Rc<Cell<bool>>,
+    signal_id: Cell<Option<SignalHandlerId>>,
```

Comment:
```suggestion
    signal_id: Option<SignalHandlerId>,
```

Doesn't have to be a cell. You only mutate it when you have a mutable reference anyway